runtime/ops/net.rs

```diff
@@ -528,3 +546,302 @@ fn op_listen(
+enum SupportedRecordType {
+  A,
+  AAAA,
+  ANAME,
+  CNAME,
+  MX,
+  PTR,
+  SRV,
+  TXT,
+}
+
+async fn op_dns_resolve(
+  state: Rc<RefCell<OpState>>,
+  args: Value,
+  _zero_copy: BufVec,
+) -> Result<Value, AnyError> {
+  fn default_port() -> u16 {
+    53
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct ResolveAddrArgs {
+    query: String,
+    record_type: SupportedRecordType,
+    options: Option<ResolveDnsOption>,
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct ResolveDnsOption {
+    name_server: Option<NameServer>,
+  }
+
+  impl Into<RecordType> for SupportedRecordType {
+    fn into(self) -> RecordType {
+      use SupportedRecordType::*;
+      match self {
+        A => RecordType::A,
+        AAAA => RecordType::AAAA,
+        ANAME => RecordType::ANAME,
+        CNAME => RecordType::CNAME,
+        MX => RecordType::MX,
+        PTR => RecordType::PTR,
+        SRV => RecordType::SRV,
+        TXT => RecordType::TXT,
+      }
+    }
+  }
+
+  #[derive(Deserialize, Clone, Copy)]
+  enum Protocol {
+    UDP,
+    TCP,
+  }
+
+  impl Default for Protocol {
+    fn default() -> Self {
+      Self::UDP
+    }
+  }
+
+  impl Into<TrustProtocol> for Protocol {
+    fn into(self) -> TrustProtocol {
+      use Protocol::*;
+      match self {
+        UDP => TrustProtocol::Udp,
+        TCP => TrustProtocol::Tcp,
+      }
+    }
+  }
+
+  #[derive(Deserialize)]
+  #[serde(rename_all = "camelCase")]
+  struct NameServer {
+    ip_addr: String,
+    #[serde(default = "default_port")]
+    port: u16,
+    #[serde(default)]
+    protocol: Protocol,
+  }
+
+  let ResolveAddrArgs {
+    query,
+    record_type,
+    options,
+  } = serde_json::from_value(args)?;
+
+  // Check for permission. If name server option is provided, use them for checking.
+  // Otherwise, checks if `allow-net` is present without any specific allowlist.
+  {
+    let perm_args = options
+      .as_ref()
+      .and_then(|opt| opt.name_server.as_ref())
+      .map(|ns| (&ns.ip_addr, ns.port));
+
+    let s = state.borrow();
+    let perm = s.borrow::<Permissions>();
+
+    if let Some((ip_addr, port)) = perm_args {
+      perm.check_net(ip_addr, port)?;
+    } else {
+      perm.check_net_all()?;
+    }
+  }
+
+  let resolver = if let Some(options) = options {
+    let conf = if let Some(name_server) = options.name_server {
+      let mut c = ResolverConfig::new();
+      c.add_name_server(NameServerConfig {
+        socket_addr: resolve_addr(&name_server.ip_addr, name_server.port)
+          .await?
+          .next()
+          .ok_or_else(|| generic_error("Invalid IP address or port"))?,
+        protocol: name_server.protocol.into(),
+        tls_dns_name: None,
+      });
+      c
+    } else {
+      let (c, _) = read_system_conf()?;
+      c
+    };
+    AsyncResolver::tokio(conf, ResolverOpts::default()).await?
+  } else {
+    AsyncResolver::tokio_from_system_conf().await?
```

Comment:
It would prefer it if this was restructured to have just a single code path, rather than two, that look up the system resolver. Easier to reason about.

As an aside: this ends up parsing `/etc/resolv.conf` for every DNS lookup. Probably not a big deal but I'd be remiss if I didn't point it out.

Reply:
You are right, I will try to refactor it.